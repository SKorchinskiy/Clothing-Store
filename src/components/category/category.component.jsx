import "./category.styles.scss";

function Category({ category: { title, imageUrl } }) {
  return (
    <div className="category-container">
      <div
        className="background-image"
        style={{
          background: `url(${imageUrl})`,
        }}
      />
      <div className="category-body-container">
        <h2>{title}</h2>
        <p>Shop Now</p>
      </div>
    </div>
  );
}

export default Category;
