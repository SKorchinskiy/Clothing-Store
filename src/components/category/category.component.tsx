import { ProductType } from "../product-item/product-item.component";
import {
  BackgroundImage,
  CategoryContainer,
  CategoryBodyContainer,
} from "./category.styles";

import { useNavigate } from "react-router-dom";

export type CategoryType = {
  id: number;
  title: string;
  imageUrl: string;
  items: ProductType[];
};

export type CategoryWithoutItems = Omit<CategoryType, "items">;

type CategoryProps = {
  category: CategoryWithoutItems;
};

function Category({ category: { title, imageUrl } }: CategoryProps) {
  const navigate = useNavigate();

  const goToCategory = () => navigate(`/category/${title}`);

  return (
    <CategoryContainer onClick={goToCategory}>
      <BackgroundImage $imageUrl={imageUrl} />
      <CategoryBodyContainer>
        <h2>{title}</h2>
        <p>Shop Now</p>
      </CategoryBodyContainer>
    </CategoryContainer>
  );
}

export default Category;
