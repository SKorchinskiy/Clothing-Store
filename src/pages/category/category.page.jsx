import {
  CategoryPreviewItems,
  CategoryPreviewContainer,
} from "./category.styles";

import { Fragment } from "react";
import { useLoaderData } from "react-router-dom";

import ProductItem from "../../components/product-item/product-item.component";

function Category() {
  const { title, items } = useLoaderData();
  return (
    <CategoryPreviewContainer>
      <h2>{title}</h2>
      <CategoryPreviewItems>
        {items.map((item) => (
          <Fragment key={item.id}>
            <ProductItem product={item}></ProductItem>
          </Fragment>
        ))}
      </CategoryPreviewItems>
    </CategoryPreviewContainer>
  );
}

export default Category;
