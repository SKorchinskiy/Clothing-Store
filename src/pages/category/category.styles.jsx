import styled from "styled-components";

export const CategoryPreviewItems = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
  column-gap: 20px;
  row-gap: 30px;
`;

export const CategoryPreviewContainer = styled.div`
  display: flex;
  flex-direction: column;

  h2 {
    margin-bottom: 20px;
    cursor: pointer;

    span {
      font-family: "Josefin Slab";
    }
  }
`;
