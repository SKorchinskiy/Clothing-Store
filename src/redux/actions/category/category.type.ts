export enum CATEGORY_ACTION_TYPES {
  FETCH_CURRENT_CATEGORY_START = "category/FETCH_CURRENT_CATEGORY_START",
  FETCH_CURRENT_CATEGORY_SUCCESS = "category/FETCH_CURRENT_CATEGORY_SUCCESS",
  FETCH_CURRENT_CATEGORY_FAILED = "category/FETCH_CURRENT_CATEGORY_FAILED",
}
