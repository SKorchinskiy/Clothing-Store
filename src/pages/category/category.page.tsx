import {
  CategoryPreviewItems,
  CategoryPreviewContainer,
} from "./category.styles";

import { Fragment, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router-dom";

import {
  selectCategoryTitle,
  selectCategoryItems,
  selectIsCategoryLoading,
} from "../../redux/selectors/category.selector";

import ProductItem from "../../components/product-item/product-item.component";
import { fetchCurrentCategoryStart } from "../../redux/actions/category/category.action";
import Loader from "../../components/loader/loader.component";

function Category() {
  const { categoryName } = useParams<string>();
  const dispatch = useDispatch();

  const title = useSelector(selectCategoryTitle);
  const items = useSelector(selectCategoryItems);
  const isLoading = useSelector(selectIsCategoryLoading);

  useEffect(() => {
    if (categoryName) {
      dispatch(fetchCurrentCategoryStart(categoryName));
    }
  }, [categoryName, dispatch]);

  return (
    <CategoryPreviewContainer>
      {isLoading ? (
        <Loader />
      ) : (
        <>
          <h2>
            <span>{title.toUpperCase()}</span>
          </h2>
          <CategoryPreviewItems>
            {items.map((item) => (
              <Fragment key={item.id}>
                <ProductItem product={item}></ProductItem>
              </Fragment>
            ))}
          </CategoryPreviewItems>
        </>
      )}
    </CategoryPreviewContainer>
  );
}

export default Category;
