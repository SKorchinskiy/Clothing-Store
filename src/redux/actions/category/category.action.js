import { createAction } from "../create-action.helper";

import { CATEGORY_ACTION_TYPES } from "./category.type";

export function fetchCurrentCategoryStart(categoryName) {
  return createAction(
    CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_START,
    categoryName
  );
}

export function fetchCurrentCategorySuccess(category) {
  return createAction(
    CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_SUCCESS,
    category
  );
}

export function fetchCurrentCategoryFailed(error) {
  return createAction(
    CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_FAILED,
    error
  );
}
