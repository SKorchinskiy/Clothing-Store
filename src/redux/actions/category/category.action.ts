import { createAction, withMatcher } from "../create-action.helper";
import { ActionWithPayload } from "../create-action.helper";
import type { Category } from "./category.type";
import { CATEGORY_ACTION_TYPES } from "./category.type";

export type FetchCategoryStart = ActionWithPayload<
  CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_START,
  string
>;

export const fetchCurrentCategoryStart = withMatcher(
  (categoryName: string): FetchCategoryStart =>
    createAction(
      CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_START,
      categoryName
    )
);

export type FetchCategorySuccess = ActionWithPayload<
  CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_SUCCESS,
  Category
>;

export const fetchCurrentCategorySuccess = withMatcher(
  (category: Category): FetchCategorySuccess =>
    createAction(CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_SUCCESS, category)
);

export type FetchCategoryFailed = ActionWithPayload<
  CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_FAILED,
  Error
>;

export const fetchCurrentCategoryFailed = withMatcher(
  (error: Error): FetchCategoryFailed =>
    createAction(CATEGORY_ACTION_TYPES.FETCH_CURRENT_CATEGORY_FAILED, error)
);
