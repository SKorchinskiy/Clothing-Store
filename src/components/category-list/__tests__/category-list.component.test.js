import "@testing-library/jest-dom";
import { screen, render } from "@testing-library/react";

import CategoryList from "../category-list.component";

import { categoriesArray } from "../../../stubs/category.stub";

jest.mock("react-router-dom", () => {
  const actualModule = jest.requireActual("react-router-dom");
  return {
    ...actualModule,
    useNavigate: () => jest.fn(),
  };
});

describe("<CategoryList /> component", () => {
  it("should render provided <Category> components", () => {
    const { asFragment } = render(
      <CategoryList categories={categoriesArray} />
    );

    const categoryElements = categoriesArray.map(({ title }) =>
      screen.getByText(title)
    );

    expect(categoryElements).toHaveLength(3);
    expect(asFragment()).toMatchSnapshot();
  });
});
