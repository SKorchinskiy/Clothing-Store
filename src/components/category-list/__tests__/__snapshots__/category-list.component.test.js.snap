// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryList /> component should render provided <Category> components 1`] = `
<DocumentFragment>
  <div
    class="sc-bcPKhP cSdPoW"
  >
    <div
      class="sc-hLclGa dfjZmB"
    >
      <div
        class="sc-dkjKgF hYJjlH"
      />
      <div
        class="sc-grXZZQ jQNJqO"
      >
        <h2>
          test category1
        </h2>
        <p>
          Shop Now
        </p>
      </div>
    </div>
    <div
      class="sc-hLclGa dfjZmB"
    >
      <div
        class="sc-dkjKgF zmKuE"
      />
      <div
        class="sc-grXZZQ jQNJqO"
      >
        <h2>
          test category2
        </h2>
        <p>
          Shop Now
        </p>
      </div>
    </div>
    <div
      class="sc-hLclGa dfjZmB"
    >
      <div
        class="sc-dkjKgF ZSDMN"
      />
      <div
        class="sc-grXZZQ jQNJqO"
      >
        <h2>
          test category3
        </h2>
        <p>
          Shop Now
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
